{
  try {
    logBasic("Starting to run...");
    while (processRow(meta,data) && !isStopped())     ;
  }
 catch (  Exception e) {
    logError("Unexpected error in '" + debug + "' : "+ e.toString());
    setErrors(1);
    stopAll();
    data.trans.stopAll();
  }
 finally {
    if (data.trans != null) {
      data.trans.waitUntilFinished();
      try {
        data.trans.endProcessing("end");
      }
 catch (      KettleException e) {
        log.logError(toString(),"Unable to log end of transformation: " + e.toString());
      }
      if (data.trans.getErrors() > 0) {
        logError("An error occurred in the sub-transformation, halting processing");
        setErrors(1);
      }
    }
    dispose(meta,data);
    logSummary();
    markStop();
  }
}
